lightly/loss/byol_loss.py

```diff
@@ -0,0 +1,189 @@
+    def flatten(t):
+        return t.reshape(t.shape[0], -1)
+
+
+class OnlineModel:
+    """
+    Online student model, encapsulates an encoder and a predictor.
+    """
+
+    def __init__(self, net, projection_size, projection_hidden_size, hidden_layer):
+        self.encoder = NetWrapper(net, projection_size, projection_hidden_size, layer=hidden_layer)
+        self.predictor = MLP(projection_size, projection_size, projection_hidden_size)
+
+    def make_prediction(self, image):
+        return self.predictor(self.encoder(image))
+
+
+class TargetModel:
+    """
+    Mean Teacher Model that takes an exponential moving average weighed by beta of past observed weights.
+    """
+
+    def __init__(self, beta, encoder, ema):
+        self.beta = beta
+        self.encoder = encoder
+        self.ema = ema
+
+    def update_with_online_encoder(self, online_encoder):
+        for current_params, ma_params in zip(online_encoder.parameters(), self.encoder.parameters()):
+            old_weight, up_weight = ma_params.data, current_params.data
+            ma_params.data = self._update_average(old_weight, up_weight)
+
+    def _update_average(self, old, new):
+        if self.ema:
+            return old * self.beta + (1 - self.beta) * new
+        else:
+            return new
+
+
+class BYOLLoss:
+    """
+    Calculates consistency loss over MSE loss of the student model and the latent representation of the
+    mean teacher (target model).
+    """
+
+    def __init__(self, params):
+        super().__init__()
+
+        self.decay = params.decay
+
+        self.augment_1, self.augment_2 = params.augment_functions
+
+        self.use_momentum = params.use_momentum
+
+        self.online = OnlineModel(params.net, params.projection_size, params.projection_hidden_size, params.layer)
+        self.target = TargetModel(params.beta, self.online.encoder)
+
+    def forward(self, x):
+        """
+
+        Args:
+            x: image
+        1. use two different augmentation function on same image
+        2. pass image through online encoder and predictor
+        3. extract projection from 'mean teacher' network
+        4. cross-wise consistency loss, sum and mean
+
+        Returns: the average consistency loss between student and
+        teacher networks for two perturbations of the same image
+
+        """
+        image_one, image_two = self.augment_1(x), self.augment_2(x)
+
+        online_pred_one = self.online.make_prediction(image_one)
+        online_pred_two = self.online.make_prediction(image_two)
+
+        with torch.no_grad():
+            target_encoder = self.target.encoder
+            target_proj_one = target_encoder(image_one).detach()
+            target_proj_two = target_encoder(image_two).detach()
+
+        loss_one = self._loss_fn(online_pred_one, target_proj_two.detach())
+        loss_two = self._loss_fn(online_pred_two, target_proj_one.detach())
+
+        loss = loss_one + loss_two
+        return loss.mean()
+
```

Comment:
Please refrain from using static methods and use module level functions instead:

```python
def _neg_cosine_loss(x, y):
   """Calculates the asymmetric negative cosine loss.

   """
   pass

class BYOLLoss(nn.Module):
...

def forward(self, x):
   """Uses neg_cosine_loss to calculate the symmetric negative cosine loss.

   """
   pass
```

Source: [Google Python Styleguide](https://google.github.io/styleguide/pyguide.html#217-function-and-method-decorators)